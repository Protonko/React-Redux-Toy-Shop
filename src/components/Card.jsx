import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Button} from './Common/Button';

export const Card = (product) => {
  const {price, sale, image, title, addToCart, countAddedProduct} = product;
  const productPrice = sale ? Math.round(price * sale) : price;
  const counterProduct = countAddedProduct ? `(${countAddedProduct})` : '';
  const classNamesPrice = classNames('card__data-price-text', {'card__data-price-text--new': sale});

  return (
    <article className="card">
      <div className="card__image">
        <img src={image} alt={title} className="card__image-pic"/>
      </div>

      <div className="card__data">
        <div className="card__section card__section--body">
          <div className="card__data-price">
            <span className={classNamesPrice}>
              ${productPrice}
            </span>
            {sale && (
              <s className="card__data-price-text card__data-price-text--old">
                ${price}
              </s>
            )}
          </div>
          <h2 className="card__data-title">
            {title}
          </h2>
        </div>
        <div className="card__section card__section--footer">
          <div className="card__data-actions">
            <ul className="card__data-actions-list list list--reset">
              <li className="card__data-actions-item">
                <Button
                  title={`Add to card ${counterProduct}`}
                  onClick={() => addToCart(product)}
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
    </article>
  );
};

Card.propTypes = {
  price: PropTypes.number,
  sale: PropTypes.number,
  countAddedProduct: PropTypes.number,
  image: PropTypes.string,
  title: PropTypes.string,
  addToCart: PropTypes.func,
};
