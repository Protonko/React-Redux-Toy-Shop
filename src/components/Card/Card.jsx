import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Button} from '../Common/Buttons/Button';

export const Card = props => {
  const {product, selectedProducts, addToCart, removeFromCart} = props;
  const {price, sale, image, title} = product;
  const isAdded = selectedProducts.find(item => item.id === product.id);
  const productPrice = sale ? Math.round(price * sale) : price;
  const titleButton = isAdded ? 'Added' : 'Add to card';
  const classNamesPrice = classNames('card__data-price-text', {'card__data-price-text--new': sale});
  const classNamesButton = classNames('button', {'button--active': isAdded});

  const handleClickCartButton = product => {
    isAdded
      ? removeFromCart(product)
      : addToCart(product);
  };

  return (
    <article className="card">
      <div className="card__image">
        <img src={image} alt={title} className="card__image-pic"/>
      </div>

      <div className="card__data">
        <div className="card__section card__section--body">
          <div className="card__data-price">
            <span className={classNamesPrice}>
              ${productPrice}
            </span>
            {sale && (
              <s className="card__data-price-text card__data-price-text--old">
                ${price}
              </s>
            )}
          </div>
          <h2 className="card__data-title">
            {title}
          </h2>
        </div>
        <div className="card__section card__section--footer">
          <div className="card__data-actions">
            <ul className="card__data-actions-list list list--reset">
              <li className="card__data-actions-item">
                <Button
                  title={titleButton}
                  classNames={classNamesButton}
                  onClick={() => handleClickCartButton(product)}
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
    </article>
  );
};

Card.propTypes = {
  product: PropTypes.object,
  selectedProducts: PropTypes.array,
  addToCart: PropTypes.func,
  removeFromCart: PropTypes.func,
};
