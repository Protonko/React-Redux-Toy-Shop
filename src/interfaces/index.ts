export interface IBaseConfig {
  baseURL: string
  headers?: Headers
}

// Products
export interface IProduct {
  category: string
  title: string
  image: string
  checked: boolean
  id: number
  price: number
  rating: number
  sale: number
}
// ./Products

// Delivery
export interface ILines {
  id: string
  description: string
  items: Array<ILinesItems>
}

export interface ILinesItems {
  id: string,
  term: string
  value: string
  marked: boolean
}

export interface IDelivery {
  id: string,
  title: string,
  description: string,
  lines: ILines
}
// ./Delivery

// Payment
export interface IPayment {
  id: string,
  title: string
  description: string
  image?: string
}
// ./Payment
