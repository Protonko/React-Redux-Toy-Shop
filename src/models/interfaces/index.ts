import {ReactComponent} from '*.svg';
import {TFilterData} from 'models/types';

export interface IBaseConfig {
  baseURL: string,
  headers?: Headers,
}

export interface IImageThumbnails {
  original: string,
  thumbnail: string,
}

export interface IError {
  message: string,
  stack?: string,
}

// Products
export interface IProduct {
  id: number,
  category: string,
  title: string,
  price: number,
  image?: string,
  sale?: number,
  rating: number,
  checked: boolean,
}

export interface IProductDetail extends  IProduct{
  description: Array<ILinesItems | void>,
  images?: Array<string>,
  about?: string,
}
// ./Product

// Socials
export interface ISocials {
  href: string,
  Icon: typeof ReactComponent,
}
// ./Socials

// Delivery
export interface ILines {
  id?: string,
  description: string,
  items: Array<ILinesItems>,
}

export interface ILinesItems {
  id?: string,
  term: string,
  value: string,
  marked: boolean,
}

export interface IDelivery {
  id?: string,
  title: string,
  description: string,
  lines: Array<ILines>,
}
// ./Delivery

// Payment
export interface IPayment {
  id: string,
  title: string,
  description: string,
  image?: string,
}
// ./Payment

// Filter
export interface IFilterData {
  id: number,
  title: string,
  code: TFilterData,
}
// ./Filter

// Navbar
export interface INavData {
  id?: number,
  href: string,
  title: string,
}
// ./Navbar

// Order
export interface IOrderCost {
  amount: number,
  total: number,
  fullPrice: number,
  sale: number,
}
// ./Order
