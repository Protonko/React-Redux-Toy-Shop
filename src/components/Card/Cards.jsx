import React from 'react';
import PropTypes from 'prop-types';
import Card from 'containers/Card';

export const Cards = ({products, selectedProducts}) => (
  <div className="cards">
    <ul className="cards__list list list--reset">
      {products.map((product, index) => (
        <li key={index} className="cards__item">
          <Card product={product} selectedProducts={selectedProducts} />
        </li>
      ))}
    </ul>
  </div>
);

Cards.propTypes = {
  products: PropTypes.object,
  selectedProducts: PropTypes.array
};
