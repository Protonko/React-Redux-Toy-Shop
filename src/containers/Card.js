import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as cartActions from '../store/actions/cart';
import {Card} from '../components/Card/Card';

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(cartActions, dispatch),
});

export default connect(null, mapDispatchToProps)(Card);
