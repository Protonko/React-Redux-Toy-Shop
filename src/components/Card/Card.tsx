// types
import {IProduct} from 'models/interfaces';
import {TITLES_BUTTON} from 'models/enums';
import {TTitleButton} from 'models/types';

import React, {useMemo, memo, FC} from 'react';
import {Link} from 'react-router-dom';
import classNames from 'classnames';
import Dotdotdot from 'react-dotdotdot';
import {checkSelectedProducts} from 'utils/checkSelectedProducts';
import {Button} from 'components/Common/Buttons/Button';
import {Rating} from 'components/Common/Rating';

export interface IProps {
  product: IProduct,
  selectedProducts: Array<IProduct>,
  addToCart: (product: IProduct) => void,
  removeFromCart: (product: IProduct) => void,
}

const CardComponent: FC<IProps> = props => {
  const {product, selectedProducts, addToCart, removeFromCart} = props;
  const {price, sale, image, title, id, rating} = product;

  const productPrice = useMemo(() => {
    return sale ? Math.round(price * sale) : price;
  }, [price, sale]);

  const isAdded = !!useMemo(() => {
    return checkSelectedProducts(selectedProducts, product);
  }, [product.id, selectedProducts]);

  const titleButton: TTitleButton = useMemo(() => {
    return isAdded ? TITLES_BUTTON.ADDED : TITLES_BUTTON.ADD;
  }, [isAdded]);

  const classNamesPrice =
    classNames('card__data-price-text', {'card__data-price-text--new': sale});

  const handleClickCartButton = (product: IProduct): void => {
    isAdded ? removeFromCart(product) : addToCart(product);
  };

  return (
    <article className="card">
      <Link to={`/product/${id}`} className="card__image">
        <img src={image} alt={title} className="card__image-pic"/>
      </Link>

      <div className="card__data">
        <div className="card__section card__section--body">
          <div className="card__data-price">
            <span className={classNamesPrice}>
              ${productPrice}
            </span>
            {sale && (
              <s className="card__data-price-text card__data-price-text--old">
                ${price}
              </s>
            )}
          </div>
          <Dotdotdot clamp={3}>
            <Link to={`/product/${id}`} className="card__data-title">{title}</Link>
          </Dotdotdot>
        </div>
        <div className="card__section card__section--footer">
          <Rating rating={rating} />
          <div className="card__data-actions">
            <ul className="card__data-actions-list list list--reset">
              <li className="card__data-actions-item">
                <Button
                  title={titleButton}
                  classNamesAdditional={[{'button--active': isAdded}]}
                  onClick={() => handleClickCartButton(product)}
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
    </article>
  );
};

export const Card = memo(CardComponent, (prevProps: any, nextProps: any): any => {
  const prevProducts = checkSelectedProducts(prevProps.selectedProducts, prevProps.product);
  const nextProducts = checkSelectedProducts(nextProps.selectedProducts, nextProps.product);

  return prevProducts === nextProducts;
});
