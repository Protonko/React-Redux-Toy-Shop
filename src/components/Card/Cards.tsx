import React, {FC} from 'react';
import {IProduct} from 'models/interfaces';
import {Card} from 'components/Card/Card';

export interface IProps {
  products: Array<IProduct>,
  selectedProducts: Array<IProduct>,
  addToCart: (product: IProduct) => void,
  removeFromCart: (product: IProduct) => void,
}

export const Cards: FC<IProps> = ({
  products,
  selectedProducts,
  addToCart,
  removeFromCart
}) => {
  const renderCard = (product: IProduct) => (
    <li key={product.id} className="cards__item">
      <Card
        product={product}
        selectedProducts={selectedProducts}
        addToCart={addToCart}
        removeFromCart={removeFromCart}
      />
    </li>
  );

  return (
    <div className="cards">
      <ul className="cards__list list list--reset">
        {products.map(renderCard)}
      </ul>
    </div>
  );
}
