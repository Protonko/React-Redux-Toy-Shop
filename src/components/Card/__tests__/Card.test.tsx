// types
import {IProps} from 'components/Card/Card';
import {IProduct} from 'models/interfaces';

import React from 'react';
import {render, shallow} from 'enzyme';
import {product, product1, selectedProducts} from 'static/__test__/card';
import {Card} from 'components/Card/Card';

const shallowCardComponent = (props: IProps) => (
  shallow(<Card {...props} />)
)

describe('Card', () => {
  let props: IProps;
  let removeFromCart: jest.Mock<boolean, [IProduct]>;
  let addToCart: jest.Mock<boolean, [IProduct]>;
  let handleClickCartButton: jest.Mock<void, [IProduct, boolean]>;

  beforeEach(() => {
    props = {
      product,
      selectedProducts,
      addToCart: () => {},
      removeFromCart: () => {},
    };
    removeFromCart = jest.fn((product) => false);
    addToCart = jest.fn((product) => true);
    handleClickCartButton = jest.fn((product, isAdded) => {
      isAdded ? removeFromCart(product) : addToCart(product);
    })
  })

  it('Should render Card component', () => {
    const component = shallowCardComponent(props);
    const wrapper = component.find('.card');
    expect(wrapper.length).toBe(1);
  });

  it('Match snapshot', () => {
    const component = shallowCardComponent(props);

    expect(component).toMatchSnapshot();
  })

  it('Should call function addToCart if array doesn\'t\n contains product', () => {
    props = {...props, selectedProducts: []};

    const isAdded = !!selectedProducts.find(
      (item: IProduct) => item.id === product1.id);

    handleClickCartButton(product, isAdded);

    expect(addToCart.mock.calls.length).toBe(1);
    expect(removeFromCart.mock.calls.length).toBe(0);
  });
});
