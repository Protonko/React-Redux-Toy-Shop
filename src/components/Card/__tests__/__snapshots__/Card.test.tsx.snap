// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Match snapshot 1`] = `
<article
  className="card"
>
  <Link
    className="card__image"
    to="/product/0"
  >
    <img
      alt="title"
      className="card__image-pic"
    />
  </Link>
  <div
    className="card__data"
  >
    <div
      className="card__section card__section--body"
    >
      <div
        className="card__data-price"
      >
        <span
          className="card__data-price-text card__data-price-text--new"
        >
          $
          9
        </span>
        <s
          className="card__data-price-text card__data-price-text--old"
        >
          $
          10
        </s>
      </div>
      <Dotdotdot
        clamp={3}
        tagName="div"
        truncationChar="â€¦"
        useNativeClamp={false}
      >
        <Link
          className="card__data-title"
          to="/product/0"
        >
          title
        </Link>
      </Dotdotdot>
    </div>
    <div
      className="card__section card__section--footer"
    >
      <Component
        rating={5}
      />
      <div
        className="card__data-actions"
      >
        <ul
          className="card__data-actions-list list list--reset"
        >
          <li
            className="card__data-actions-item"
          >
            <Component
              classNamesAdditional={
                Array [
                  Object {
                    "button--active": true,
                  },
                ]
              }
              onClick={[Function]}
              title="Added"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>
</article>
`;
